Specify how function pointers will handle context registers (#309)
dotnet/designs: proposed/swift-interop.md

#### Calling Conventions

> Swift has two calling conventions at its core, the "Swift" calling convention, and the "SwiftAsync" calling convention. We will begin by focusing on the "Swift" calling convention, as it is the most common. The Swift calling convention has a few elements that we need to contend with. Primarily, it allows passing arguments in more registers. Additionally, it has two dedicated registers for specific components of the calling convention, the self register and the error register. The runtime support for this calling convention must support all of these features. The additional registers per argument is relatively easy to support, as each of our compilers must already have some support for this concept to run on Linux or Apple platforms today. We have a few options for the remaining two features we need to support. The `CallConvSwift` modifier can be applied in all contexts where other existing calling convention modifiers can be used. This means it can be used with UnmanagedCallConvAttribute for P/Invoke (`[UnmanagedCallConv(CallConvs = new Type[] { typeof(CallConvSwift) })]`), UnmanagedCallersOnlyAttribute.CallConv for reverse P/Invoke (`[UnmanagedCallersOnly(CallConvs = new Type[] { typeof(CallConvSwift) })]`), and with unmanaged function pointers (`delegate* unmanaged[Swift]<>`).

Reviewer: ```suggestion
Swift has two calling conventions at its core, the "Swift" calling convention, and the "SwiftAsync" calling convention. We will begin by focusing on the "Swift" calling convention, as it is the most common. The Swift calling convention has a few elements that we need to contend with. Primarily, it allows passing arguments in more registers. Additionally, it has two dedicated registers for specific components of the calling convention, the self register and the error register. The runtime support for this calling convention must support all of these features. The additional registers per argument is relatively easy to support, as each of our compilers must already have some support for this concept to run on Linux or Apple platforms today. We have a few options for the remaining two features we need to support. The `CallConvSwift` modifier can be applied in all contexts where other existing calling convention modifiers can be used. This means it can be used with [`UnmanagedCallConvAttribute`](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.unmanagedcallconvattribute) for P/Invoke (`[UnmanagedCallConv(CallConvs = new Type[] { typeof(CallConvSwift) })]`), [`UnmanagedCallersOnlyAttribute.CallConv`](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.unmanagedcallersonlyattribute.callconvs) for reverse P/Invoke (`[UnmanagedCallersOnly(CallConvs = new Type[] { typeof(CallConvSwift) })]`), and with unmanaged function pointers (`delegate* unmanaged[Swift]<>`).
```

1. Use specially-typed argument with a type, `SwiftSelf`, to represent which parameter should go into the self register.

> We will provide a `SwiftSelf` type to specify "this argument goes in the self register". Specifying the type twice in a signature would generate an `InvalidProgramException`. This would allow the `self` argument to be specified anywhere in the argument list. Alternatively, many sections of the Swift ABI documentation, as well as the Clang docs refer to this parameter as the "context" argument instead of the "self" argument, so an alternative name could be `SwiftContext`. Method annotated with `UnmanagedCallersOnly` will be treated as Swift-like functions capable of handling context registers. The value stored in the self register will be loaded into the `SwiftSelf` argument.

Reviewer 2: ```suggestion
We will provide a `SwiftSelf` type to specify "this argument goes in the self register". Specifying the type twice in a signature would generate an `InvalidProgramException`. This would allow the `self` argument to be specified anywhere in the argument list. Alternatively, many sections of the Swift ABI documentation, as well as the Clang docs refer to this parameter as the "context" argument instead of the "self" argument, so an alternative name could be `SwiftContext`. Methods annotated with `UnmanagedCallersOnly` will be treated as Swift-like functions capable of handling context registers. The value stored in the self register will be loaded into the `SwiftSelf` argument.
```

> We will provide a `SwiftSelf` type to specify "this argument goes in the self register". Specifying the type twice in a signature would generate an `InvalidProgramException`. This would allow the `self` argument to be specified anywhere in the argument list. Alternatively, many sections of the Swift ABI documentation, as well as the Clang docs refer to this parameter as the "context" argument instead of the "self" argument, so an alternative name could be `SwiftContext`. Methods annotated with `UnmanagedCallersOnly` will be treated as Swift-like functions capable of handling context registers. The value stored in the self register will be loaded into the `SwiftSelf` argument.

Reviewer: ```suggestion
We will provide a `SwiftSelf` type to specify "this argument goes in the self register". Specifying the type twice in a signature would generate an `InvalidProgramException`. This would allow the `self` argument to be specified anywhere in the argument list. Alternatively, many sections of the Swift ABI documentation, as well as the Clang docs refer to this parameter as the "context" argument instead of the "self" argument, so an alternative name could be `SwiftContext`. Methods annotated with `UnmanagedCallersOnly` will be capable of handling the self register; the value stored in the self register will be loaded into the `SwiftSelf` argument.
```

> We will provide a `SwiftSelf` type to specify "this argument goes in the self register". Specifying the type twice in a signature would generate an `InvalidProgramException`. This would allow the `self` argument to be specified anywhere in the argument list. Alternatively, many sections of the Swift ABI documentation, as well as the Clang docs refer to this parameter as the "context" argument instead of the "self" argument, so an alternative name could be `SwiftContext`. Function pointers annotated with `UnmanagedCallersOnly` will be treated as Swift-like functions capable of handling context registers. The value stored in the self register will be loaded into the `SwiftSelf` argument.

Reviewer 2: ```suggestion
We will provide a `SwiftSelf` type to specify "this argument goes in the self register". Specifying the type twice in a signature would generate an `InvalidProgramException`. This would allow the `self` argument to be specified anywhere in the argument list. Alternatively, many sections of the Swift ABI documentation, as well as the Clang docs refer to this parameter as the "context" argument instead of the "self" argument, so an alternative name could be `SwiftContext`. Method annotated with `UnmanagedCallersOnly` will be treated as Swift-like functions capable of handling context registers. The value stored in the self register will be loaded into the `SwiftSelf` argument.
```
Function pointers cannot be annotate with `UnmanagedCallersOnly`.

Author: Yes, function pointer is a handle, thanks for the suggestion.

> We will provide a `SwiftSelf` type to specify "this argument goes in the self register". Specifying the type twice in a signature would generate an `InvalidProgramException`. This would allow the `self` argument to be specified anywhere in the argument list. Alternatively, many sections of the Swift ABI documentation, as well as the Clang docs refer to this parameter as the "context" argument instead of the "self" argument, so an alternative name could be `SwiftContext`. The `UnmanagedCallersOnly` leverages the existing proposal and treats delegates annotated with `UnmanagedCallersOnly` as Swift-like functions capable of handling context registers. In the prolog of the delegate, the self register value will be loaded into the `SwiftSelf` argument.

Reviewer 2: C# overloaded the delegate keyword for function pointers. We should refer to function pointers as function pointers to avoid confusion.

1. Use a special type named something like `SwiftError*` to indicate the error parameter

> This approach expresses that the to-be-called Swift function uses the Error Register in the signature and they both require signature manipulation in the JIT/AOT compilers. Like with `SwiftSelf`, we would throw an `InvalidProgramException` for a signature with multiple `SwiftError` parameters. We use a pointer-to-`SwiftError` type to indicate that the error register is a by-ref/out parameter. We don't use managed pointers as our modern JITs can reason about unmanaged pointers well enough that we do not end up losing any performance taking this route. The `UnmanagedCallersOnly` implementation will require a decent amount of JIT work to emulate a local variable for the register value, but we have prior art in the Clang implementation of the Swift error register that we can fall back on. Delegates annotated with `UnmanagedCallersOnly` will store the `SwiftError*` value in the error register in their epilog.

Reviewer 2: (Same as above.)

It is codegen implementation detail whether it is going to store the error registers somewhere in the prolog; or whether it is going to keep it enregistered throughout the method. I expect that RyuJIT may be able to do the later some of the time.
